src/Catel.MVVM/Catel.MVVM.Shared/Services/UIVisualizerService.wpf.cs

```diff
@@ -130,32 +130,34 @@ protected virtual void HandleCloseSubscription(object window, object data, Event
         {
             Argument.InheritsFrom<System.Windows.Window>("window", window);
 
-            return Task.Run(() => {
-                    if (showModal)
-                    {
-                        // Child window does not have a ShowDialog, so not null is allowed
-                        bool? result = null;
+            var tsc = new TaskCompletionSource<bool?>();
+            
+                if (showModal)
+                {
+                    // Child window does not have a ShowDialog, so not null is allowed
+                    bool? result;
 
-                        window.Dispatcher.BeginInvoke(() =>
+                    window.Dispatcher.BeginInvoke(() =>
+                        {
+                            // Safety net to prevent crashes when this is the main window
+                            try
+                            {
+                                result = ((System.Windows.Window)window).ShowDialog();
+                                tsc.SetResult(result);
+                            }
+                            catch (Exception ex)
                             {
-                                // Safety net to prevent crashes when this is the main window
-                                try
-                                {
-                                    result = ((System.Windows.Window)window).ShowDialog();
-                                }
-                                catch (Exception ex)
-                                {
-                                    Log.Warning(ex, $"An error occurred while showing window '{window.GetType().GetSafeFullName(true)}'");
-                                }
-                            });
-
-                        return result;
-                    }
-
-                    window.Dispatcher.BeginInvoke(() => ((System.Windows.Window)window).Show());
+                                Log.Warning(ex, $"An error occurred while showing window '{window.GetType().GetSafeFullName(true)}'");
+                                tsc.SetResult(null);
+                            }
+                        });
 
-                    return null;
-                });
+                    return tsc.Task;
+                }
+
+                window.Dispatcher.BeginInvoke(() => ((System.Windows.Window)window).Show());
+                tsc.SetResult(null);
```

Comment:
This tsc.SetResult doesn't really make sense. We need to subscribe to the close event of the window (even when using Show instead of ShowDialog). Then users can choose whether they want to await the result or not (but that's up to the user).